python/ray/data/tests/test_autoscaler.py

```diff
@@ -82,22 +84,47 @@ def patch(mock, attr, value, is_method=True):
         yield
         setattr(mock, attr, original)
 
+    ExpectedReason = Union[str, Callable[[str], bool], None]
+
     def assert_autoscaling_action(
-        *, delta: int, expected_reason: Optional[str], force: bool = False
+        *, delta: int, expected_reason: ExpectedReason, force: bool = False
     ):
         nonlocal actor_pool, op, op_state
 
-        assert autoscaler._derive_target_scaling_config(
+        request = autoscaler._derive_target_scaling_config(
             actor_pool=actor_pool,
             op=op,
             op_state=op_state,
-        ) == ActorPoolScalingRequest(delta=delta, force=force, reason=expected_reason)
+        )
+
+        assert request.delta == delta
+        assert request.force == force
+
+        if callable(expected_reason):
+            assert expected_reason(
+                request.reason
+            ), f"Unexpected reason: {request.reason}"
+        else:
+            assert request.reason == expected_reason
+
+    def calculate_plan_delta(util: float, current_size: int, threshold: float) -> int:
+        """Calculate plan_delta based on utilization formula."""
+        return (
+            math.ceil(current_size * (util / threshold - 1))
+            if current_size > 0 and util > 0
+            else 1
+        )
```

Comment:
Re-implementing logic in tests hurts readability and increases the risk of introducing buggy tests (e.g., if the code is buggy, and the tests just duplicate it). Could you separate the `max_upscaling_delta` tests into its separate test function(s) and explicitly testing against an expected delta?